ssl/quic/quic_ackm.c

```diff
@@ -112,20 +110,13 @@ tx_pkt_history_add_actual(struct tx_pkt_history_st *h,
         return 0;
 
     /* Should not already be in a list. */
-    if (!ossl_assert(pkt->next == NULL && pkt->prev == NULL))
+    if (!ossl_assert(ossl_list_tx_history_next(pkt) == NULL
```

Comment:
Did you consider us changing to a thread safe implementation at some point?  Which I see as possible / even likely.  It's certainly something I'd like to do at some point.  Either using locks or a lock free version?  Any refactor at that point would be painful.

I'm strongly opposed to allowing direct access to the internal fields.  There is zero performance gain by doing so & it impact readability IMO.  It restricts the data structure's future unnecessarily.

I can cherry pick implementations that **do** use accessor functions.  I don't consider that a valid argument.

Reply:
>Restricting access to internals is the fundamental premise of data type abstraction.

Sure, but I'm not *at all* interested in data type abstraction in this case. The benefit of this code as I see it is in factoring commonly used code, not in creating an abstract data type.

I don't see changing this code to be thread safe in the future to be a remotely sane proposition. In fact I'm happy to outright rule out any such change in the future, and frankly I can't see the benefit. No code which uses this list API is going to magically become thread safe just because this list code becomes thread safe. You aren't going to escape the need to review and refactor a specific body of code using this API in order to accomplish threaded operation. Moreover, making the list type thread safe implies doing fine-grained locking at the level of the list data structure, which is inefficient as only the user of the list API is in an appropriate position to determine the correct unit of locking, and may already hold relevant locks.

Use of accessors also precludes taking a pointer to the next/prev fields, which precludes some useful techniques, for example in the case of singly linked lists (see quic_ackm.c:719).

>I can just as easily cherry pick projects that do use accessor functions. Proof by example is a classic logical fallacy and is not a compelling argument.

My comment is a request for empirical evidence, not some invocation of a "fallacy". If you have empirical evidence of a C project which used accessors for something as fundamental as a doubly linked list, *and then gained major advantage from this decision*, I'd find that to be information worth taking into account.

Reply:
How does one demonstrate _gained major advantage_?

You examples certainly don't do this.